common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
Anyway, even if it can be set somehow, it's possible that it doesn't matter for a material checkout or clone on the server, but it is sent to the agent and it doesn't the right thing while preparing for a job run. In that sense, it is one material on the server, but a slightly different one on the agent. As long as the final revision of the checkout is ok, the agent shouldn't care.

That's what I was wondering whether we could do for shallow clone. Ignoring the shallowClone option on the server side and doing whatever is needed to make polling work (user's don't care about that clone. It's internal to the Go Server). On the agent side, do what the users asked for.